src/cart/src/cart/crt_group.c

```diff
@@ -359,41 +335,39 @@ grp_li_uri_set(struct crt_lookup_item *li, int tag, const char *uri)
 		D_INIT_LIST_HEAD(&ui->ui_link);
 		ui->ui_ref = 0;
 		ui->ui_initialized = 1;
-
-		rc = D_MUTEX_INIT(&ui->ui_mutex, NULL);
-		if (rc != 0) {
+		ui->ui_rank = li->li_rank;
+		D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
+		if (!ui->ui_uri[tag]) {
+			D_ERROR("Failed to duplicate an uri\n");
 			D_FREE_PTR(ui);
-			D_GOTO(exit, rc);
+			D_GOTO(exit, rc = -DER_NOMEM);
 		}
 
-		ui->ui_rank = li->li_rank;
-
 		rc = d_hash_rec_insert(&grp_priv->gp_uri_lookup_cache,
 				&rank, sizeof(rank),
 				&ui->ui_link,
 				true /* exclusive */);
 		if (rc != 0) {
 			D_ERROR("Entry already present\n");
-			D_MUTEX_DESTROY(&ui->ui_mutex);
+			D_FREE(ui->ui_uri[tag]);
 			D_FREE_PTR(ui);
 			D_GOTO(exit, rc);
 		}
-		D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
-
-		if (!ui->ui_uri[tag]) {
-			d_hash_rec_delete(&grp_priv->gp_uri_lookup_cache,
-					&rank, sizeof(d_rank_t));
-			D_GOTO(exit, rc = -DER_NOMEM);
-		}
 	} else {
 		ui = crt_ui_link2ptr(rlink);
-		if (!ui->ui_uri[tag]) {
-			D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
-		}
+		if (atomic_load_consume(&ui->ui_uri[tag]) == NULL) {
```

Comment:
I'm not sure what this block is doing, why is it doing a atomic read, followed by an atomic swap?

Would a single call to atomic_compare_exchange be adequate here?

Reply:
probably this is over care.